File: 22-Binary-Trees-Miscellaneous/java/06-FloodFill.java
import java.util.*;

class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        int originalColor = image[sr][sc];
        if (originalColor == color) return image;

        int rows = image.length, cols = image[0].length;
        int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{sr, sc});
        image[sr][sc] = color;

        while (!queue.isEmpty()) {
            int[] pixel = queue.poll();
            for (int[] dir : directions) {
                int newRow = pixel[0] + dir[0];
                int newCol = pixel[1] + dir[1];

                if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols
                        && image[newRow][newCol] == originalColor) {
                    image[newRow][newCol] = color;
                    queue.offer(new int[]{newRow, newCol});
                }
            }
        }
        return image;
    }
}

File: 22-Binary-Trees-Miscellaneous/python/06_flood_fill.py
from collections import deque
from typing import List

class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        original_color = image[sr][sc]
        if original_color == color:
            return image

        rows, cols = len(image), len(image[0])
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]

        queue = deque([(sr, sc)])
        image[sr][sc] = color

        while queue:
            r, c = queue.popleft()
            for dr, dc in directions:
                new_row, new_col = r + dr, c + dc
                if 0 <= new_row < rows and 0 <= new_col < cols and image[new_row][new_col] == original_color:
                    image[new_row][new_col] = color
                    queue.append((new_row, new_col))

        return image
